File: web/src/lib/rice.ts
// SPDX-License-Identifier: AGPL-3.0-or-later

// Mirrors internal/artifact/rice.go — keep the formula, constraints, and
// validation messages in lockstep with the Go implementation (single source
// of truth per requirement §22).

export interface RiceComponents {
  rice_reach?: number | null
  rice_impact?: number | null
  rice_confidence?: number | null
  rice_effort?: number | null
}

export type RiceField = 'rice_reach' | 'rice_impact' | 'rice_confidence' | 'rice_effort'

/** Editor pre-fill only, applied when opening the editor on an item with no RICE fields set. */
export const RICE_DEFAULTS = {
  reach: 100,
  impact: 0.25,
  confidence: 25,
  effort: 1,
} as const

export const IMPACT_TIERS = [0.25, 0.5, 1, 2, 3] as const

/**
 * riceScore derives the RICE score from c's four components:
 * (reach × impact × (confidence/100)) / effort.
 * Returns null ("N/A") unless all four components are present and each
 * satisfies its constraint — mirrors internal/artifact/rice.go RiceScore.
 */
export function riceScore(c: RiceComponents | null | undefined): number | null {
  if (!c) return null
  const { rice_reach: reach, rice_impact: impact, rice_confidence: confidence, rice_effort: effort } = c
  if (reach == null || impact == null || confidence == null || effort == null) return null
  if ([reach, impact, confidence, effort].some((v) => Number.isNaN(v))) return null
  if (reach < 0 || impact < 0) return null
  if (confidence < 0 || confidence > 100) return null
  if (effort <= 0) return null
  return (reach * impact * (confidence / 100)) / effort
}

export function formatRice(v: number | null | undefined): string {
  return v == null ? 'N/A' : v.toFixed(2)
}

/**
 * validateRiceComponent mirrors internal/artifact/rice.go ValidateRiceComponent.
 * A null/undefined value is always valid (the unset state) — it returns a
 * field-level message only when a present value violates its constraint.
 */
export function validateRiceComponent(field: RiceField, value: number | null | undefined): string | null {
  if (value == null) return null
  if (Number.isNaN(value)) return 'must be a number'
  switch (field) {
    case 'rice_reach':
    case 'rice_impact':
      return value < 0 ? 'must be >= 0' : null
    case 'rice_confidence':
      return value < 0 || value > 100 ? 'must be between 0 and 100' : null
    case 'rice_effort':
      return value <= 0 ? 'must be > 0' : null
  }
}

File: internal/artifact/rice.go
// SPDX-License-Identifier: AGPL-3.0-or-later

package artifact

import (
	"fmt"
	"math"
)

// RiceValidationError reports a field-level RICE validation failure.
type RiceValidationError struct {
	Field   string
	Message string
}

func (e *RiceValidationError) Error() string {
	return fmt.Sprintf("%s: %s", e.Field, e.Message)
}

// RiceScore derives the RICE score from fm's four components:
// (reach × impact × (confidence/100)) / effort.
// ok is true only when all four components are set and each satisfies its
// constraint (see ValidateRiceComponent); otherwise ok is false ("N/A") and
// score is 0. The returned score is never Inf or NaN.
func RiceScore(fm Frontmatter) (score float64, ok bool) {
	if fm.RiceReach == nil || fm.RiceImpact == nil || fm.RiceConfidence == nil || fm.RiceEffort == nil {
		return 0, false
	}
	if ValidateRice(fm) != nil {
		return 0, false
	}
	return (*fm.RiceReach * *fm.RiceImpact * (*fm.RiceConfidence / 100)) / *fm.RiceEffort, true
}

// ValidateRiceComponent validates a single RICE component value against its
// field-specific constraint. A nil value is always valid — it represents the
// unset state, not an error. field must be one of "rice_reach", "rice_impact",
// "rice_confidence", "rice_effort".
func ValidateRiceComponent(field string, v *float64) error {
	if v == nil {
		return nil
	}
	switch field {
	case "rice_reach":
		if *v < 0 {
			return &RiceValidationError{Field: field, Message: "must be >= 0"}
		}
	case "rice_impact":
		if *v < 0 {
			return &RiceValidationError{Field: field, Message: "must be >= 0"}
		}
	case "rice_confidence":
		if *v < 0 || *v > 100 {
			return &RiceValidationError{Field: field, Message: "must be between 0 and 100"}
		}
	case "rice_effort":
		if *v <= 0 {
			return &RiceValidationError{Field: field, Message: "must be > 0"}
		}
	default:
		return &RiceValidationError{Field: field, Message: "unknown RICE field"}
	}
	return nil
}

// ValidateRice validates every RICE component present on fm, returning the
// first field-level error encountered, or nil when all present components
// satisfy their constraints. Components left unset (nil) are not required.
func ValidateRice(fm Frontmatter) error {
	if err := ValidateRiceComponent("rice_reach", fm.RiceReach); err != nil {
		return err
	}
	if err := ValidateRiceComponent("rice_impact", fm.RiceImpact); err != nil {
		return err
	}
	if err := ValidateRiceComponent("rice_confidence", fm.RiceConfidence); err != nil {
		return err
	}
	if err := ValidateRiceComponent("rice_effort", fm.RiceEffort); err != nil {
		return err
	}
	return nil
}

// RoundRice rounds score to 2 decimal places for display. Sorting and storage
// use the unrounded value returned by RiceScore.
func RoundRice(score float64) float64 {
	return math.Round(score*100) / 100
}
